documentation/specs/proposed/Wasm-tasks.md

```diff
@@ -0,0 +1,344 @@
+2. Write an `ITaskFactory` and a supporting `ITask` classes that when passed a `.wasm` file implementing required functions runs it as an MSBuild task
+3. Demos/examples
+
+### Prototype features
+Prototypes are implemented in [https://github.com/JanProvaznik/MSBuildWasm](https://github.com/JanProvaznik/MSBuildWasm)
+- [ ] WasmExec class taking a .wasm file as a parameter - just runs the file with Wasmtime 
+    - nudges the user to parametrize access to resources
+- [ ] WasmTask - creating tasks from .wasm files
+    - [x] Specification for what should this .wasm file export and how it will be ran
+    - [ ] ITaskFactory that let's msbuild use task parameters defined inside the .wasm module
+- [ ] Rust example
+#### User Experience
+1. The user Writes a task in Rust based on the template.
+2. The user adds the task to their .proj file and it runs and logs as if it were a C# task. 
+```xml
+<UsingTask TaskName="FancyWasmTask"
+           AssemblyFile="path/MSBuildWasm.dll"
+           TaskFactory="WasmTaskFactory">
+  <Task>
+    <WasmModule>compiled_task_implementation.wasm</WasmModule>
+  </Task>
+</UsingTask>
+
+<Target Name="name">
+<FancyWasmTask Param="..." Param2="asdf">
+<Output .../>
+</FancyWasiTask>
+</Target>
+```
+
+### Advanced features
+- [ ] ~~.NET example~~ (WASIp1 will not be supported in .NET)
+- [ ] integrating pipeline for creating Wasm/WASI tasks from code in Rust 
+    - [ ] investigate integrating tools compiling languages to Wasm/WASI
+    - On task level
+        - [ ] RustTaskFactory
+        - exploring other languages (Go, C/C++, Zig)
+- [x] investigate running an arbitrary .NET task distributed as a dll in the WASI sandbox (👀 Mono runtime)
+    - Due to implementing WasmTasks with WASIp1, it will not compatible with .NET runtime effort to support WASIp2
+
+
+## Design
+### diagram
+
+```mermaid
+flowchart TD
+    A[MSBuild] -->|Evaluation| B[WasmTaskFactory]
+    A -->|Target execution| C[TaskExecutionHost]
+    C -->|instantiate and\n set parameters from XML| D[WasmTask]
+    H[Rust/C#/Go] -->|"compile using wasi-sdk"| G
+    D -->|gather output \nfor use in other tasks| C 
+    D -->|execute| E[wasmtime-dotnet]
+    E <--> F[Wasmtime]
+
+    B -->|Create Type for a specific WasmTask| D
+    B -->|read what the task expects as parameters| E
+    B -->|save path to task parameters| G[.wasm module]
+    E -->|read output from task stdout| D
+    %%B, C, D%%
+    style B fill:#ffff00
+    style C fill:#ffff00
+    style D fill:#ffff00
+```
+C# classes are yellow.
+
+
+### Wasm/WASI communication with MSBuild
+Without WASIp2 WIT (not implemented in wasmtime-dotnet), the only data type that an be a Wasm function parameter and output is a number. Tasks have parameters which are of the following types: string, bool, [ITaskItem](https://github.com/dotnet/msbuild/blob/main/src/Framework/ITaskItem.cs) (basically a string dict), and arrays of these types.
```

Comment:
Are there plans to add it? Link?

Reply:
wasmtime-dotnet does not have a roadmap now afaik. I had made an issue for it, and now added it to this spec: https://github.com/bytecodealliance/wasmtime-dotnet/issues/324